src/filters/compression.rs

```diff
@@ -0,0 +1,263 @@
+            CompressionAlgo::DEFLATE => HeaderValue::from_static("deflate"),
+            CompressionAlgo::GZIP => HeaderValue::from_static("gzip"),
+        }
+    }
+}
+
+/// Compression
+#[derive(Clone, Copy, Debug)]
+pub struct Compression<F> {
+    func: F,
+}
+
+// TODO: The implementation of `gzip()`, `deflate()`, and `brotli()` could be replaced with
+// generics or a macro
+
+/// Create a wrapping filter that compresses the Body of a [`Response`](crate::reply::Response)
+/// using gzip, adding `content-encoding: gzip` to the Response's [`HeaderMap`](hyper::HeaderMap)
+///
+/// # Example
+///
+/// ```
+/// use warp::Filter;
+///
+/// let route = warp::get()
+///     .and(warp::path::end())
+///     .and(warp::fs::file("./README.md"))
+///     .with(warp::compression::gzip());
+/// ```
+pub fn gzip() -> Compression<impl Fn(CompressionProps) -> Response + Copy> {
+    let func = move |mut props: CompressionProps| {
+        let body = Body::wrap_stream(GzipEncoder::new(props.body));
+        props
+            .head
+            .headers
+            .append(CONTENT_ENCODING, CompressionAlgo::GZIP.into());
+        Response::from_parts(props.head, body)
+    };
+    Compression { func }
+}
+
+/// Create a wrapping filter that compresses the Body of a [`Response`](crate::reply::Response)
+/// using deflate, adding `content-encoding: deflate` to the Response's [`HeaderMap`](hyper::HeaderMap)
+///
+/// # Example
+///
+/// ```
+/// use warp::Filter;
+///
+/// let route = warp::get()
+///     .and(warp::path::end())
+///     .and(warp::fs::file("./README.md"))
+///     .with(warp::compression::deflate());
+/// ```
+pub fn deflate() -> Compression<impl Fn(CompressionProps) -> Response + Copy> {
+    let func = move |mut props: CompressionProps| {
+        let body = Body::wrap_stream(DeflateEncoder::new(props.body));
+        props
+            .head
+            .headers
+            .append(CONTENT_ENCODING, CompressionAlgo::DEFLATE.into());
+        Response::from_parts(props.head, body)
+    };
+    Compression { func }
+}
+
+/// Create a wrapping filter that compresses the Body of a [`Response`](crate::reply::Response)
+/// using brotli, adding `content-encoding: br` to the Response's [`HeaderMap`](hyper::HeaderMap)
+///
+/// # Example
+///
+/// ```
+/// use warp::Filter;
+///
+/// let route = warp::get()
+///     .and(warp::path::end())
+///     .and(warp::fs::file("./README.md"))
+///     .with(warp::compression::brotli());
+/// ```
+pub fn brotli() -> Compression<impl Fn(CompressionProps) -> Response + Copy> {
+    let func = move |mut props: CompressionProps| {
+        let body = Body::wrap_stream(BrotliEncoder::new(props.body));
+        props
+            .head
+            .headers
+            .append(CONTENT_ENCODING, CompressionAlgo::BR.into());
+        Response::from_parts(props.head, body)
+    };
+    Compression { func }
+}
+
+/// Compression Props
+#[derive(Debug)]
+pub struct CompressionProps {
```

Comment:
Does this need to be public?

Reply:
It does, because CompressionProps is used in the internal module, but also is exposed externally via the return type of the filter functions, e.g. `gzip() -> Compression<impl Fn(CompressionProps) -> Response + Copy>`

Reply:
Hm, what if this were moved to the `mod internal` then, since the argument to the closure is just an implementation detail...

Reply:
I moved `CompressionProps` into `mod internal` (and `CompressableBody` because it seemed to make sense), but this gets into a part of Rust I don't know about, and thats module visibility.

`CompressionProps` still needed to be public because of the return type of the `gzip() -> ...` and such, but does `CompressionProps` being in a private module protect its visibility?
Also I totally could have missed the point here, and maybe what I did makes no difference?